### PR TITLE
feat: add verify simple justification

### PR DESCRIPTION
**Pull request by xavierdmello**
Add library function to verify a simple justification on a block from the specified authority set

### CODE REVIEW

File: primitives/src/lib.rs [hunk 1]
```diff
@@ -1,5 +1,5 @@
 use ed25519_dalek::{Signature, Verifier, VerifyingKey};
-
+use ethers::types::H256;
```
Comment (ratankaliani) [inline primitives/src/lib.rs]: You should use alloy_sol_types::B256 here instead. ethers-rs won't compile inside of the program, so you shouldn't put it in the primitives folder.

File: .gitignore [hunk 2]
```diff
@@ -23,4 +23,6 @@ pgo-data.profdata
 .env
 
 # vscode debug config
-.vscode/launch.json
\ No newline at end of file
+.vscode/launch.json
+
+#
```
Comment (ratankaliani) [inline .gitignore]: nit: remove

File: primitives/src/lib.rs [hunk 3]
```diff
@@ -1,16 +1,109 @@
...
 pub fn verify_signature(pubkey_bytes: &[u8; 32], signed_message: &[u8], signature: &[u8; 64]) {
-    let pubkey = VerifyingKey::from_bytes(pubkey_bytes).unwrap();
+    let pubkey: VerifyingKey = VerifyingKey::from_bytes(pubkey_bytes).unwrap();
     let verified = pubkey.verify(signed_message, &Signature::from_bytes(signature));
     if verified.is_err() {
         panic!("Signature is not valid");
     }
 }
 
+// Verify a simple justification on a block from the specified authority set
```
Comment (ratankaliani) [inline primitives/src/lib.rs]: For function desc comments you should use `///` which is a rustdoc comment for functions. You can see an example of this here: https://github.com/ratankaliani/sp1-vectorx/blob/main/header-range/src/main.rs#L80. 

Additionally, always put a period at the end of comments.

File: primitives/src/lib.rs [hunk 4]
```diff
@@ -1,16 +1,109 @@
...
+pub fn verify_simple_justification(justification: CircuitJustification, authority_set_id: u64, authority_set_hash: Vec<u8>) {
+    // 1. Justification is untrusted and must be linked to verified authority set hash
+    let computed_authority_set_commitment = compute_authority_set_commitment(justification.num_authorities, justification.pubkeys.clone());
+   
+    // Verify the authority set commitment is valid.
+    assert_eq!(computed_authority_set_commitment, authority_set_hash);
+
+    // 2. Check encoding of precommit mesage
+    // a) decode precommit
+    // b) check that values from decoded precommit match passes in block number, block hash, and authority_set_id
```
Comment (ratankaliani) [inline primitives/src/lib.rs]: nit: Check that values from decoded precommit match passed in block number, block hash and authority_set_id.

File: primitives/src/lib.rs [hunk 5]
```diff
@@ -1,16 +1,109 @@
...
+    let (signed_block_hash, signed_block_number, _, signed_authority_set_id) = decode_precommit(justification.signed_message.clone());
+    assert_eq!(signed_block_hash, justification.block_hash);
+    assert_eq!(signed_block_number, justification.block_number);
+    assert_eq!(signed_authority_set_id, authority_set_id);
+
+    // 3. Check that the signed message is signed by the correct authority
+    assert_eq!(justification.pubkeys.len(), justification.signatures.len());
```
Comment (ratankaliani) [inline primitives/src/lib.rs]: What does this check do?

File: .gitignore (hunk 2 above)
Comment (xavierdmello) [reply .gitignore]: isnt this good to have in the gitignore incase someone makes a debug profile so it doesnt get added (like it did in the last pr)

File: primitives/src/lib.rs [hunk 6]
```diff
@@ -1,16 +1,107 @@
...
+    // a) Decode precommit
+    // b) Check that values from decoded precommit match passed in block number, block hash and authority_set_id.
+    let (signed_block_hash, signed_block_number, _, signed_authority_set_id) = decode_precommit(justification.signed_message.clone());
+    assert_eq!(signed_block_hash, justification.block_hash);
+    assert_eq!(signed_block_number, justification.block_number);
+    assert_eq!(signed_authority_set_id, authority_set_id);
+
+    // 3. Check that the signed message is signed by the correct authority
+    // Two is added to account for rounding in case the number of authorities is not divisible by 3
+    let threshold = (2 * justification.pubkeys.len() + 2) / 3;
```
Comment (ratankaliani) [inline primitives/src/lib.rs]: Use `ceil`

File: rotate/src/main.rs [hunk 7]
```diff
@@ -19,28 +17,14 @@ pub fn main() {
 
     // Compute new authority set hash & convert it from binary to bytes32 for the blockchain
     let new_authority_set_hash: Vec<u8> =
-        compute_authority_set_commitment(justification.num_authorities, justification.pubkeys);
+        compute_authority_set_commitment(justification.num_authorities, justification.pubkeys.clone());
     let new_authority_set_hash_bytes32: [u8; 32] = new_authority_set_hash
         .try_into()
         .expect("Failed to convert hash to bytes32");
+
+    // Verify simple justification, panics if fails
```
Comment (ratankaliani) [inline rotate/src/main.rs]: nit: Remove this line.

File: script/src/bin/rotate.rs [hunk 8]
```diff
@@ -19,7 +19,7 @@ async fn main() {
     // Fetch the authority set hash for the specified authority set id.
     // TODO: In the future, this will be read from the contract, along with the epoch end block number.
     let authority_set_hash = fetcher
-        .compute_authority_set_hash(epoch_end_block - 1)
```
Comment (ratankaliani) [inline script/src/bin/rotate.rs:22]: This line should keep the - 1. In the future we can refactor this, but for now, `compute_authority_set_hash` will work the way it currently it does.

File: script/src/input.rs [hunk 9]
```diff
@@ -347,13 +319,16 @@ impl RpcDataFetcher {
...
+
         CircuitJustification {
             signed_message,
             authority_set_id,
             current_authority_set_hash,
             pubkeys: authorities.clone(),
             signatures,
             num_authorities,
+            block_number,
+            block_hash: self.get_block_hash(block_number).await.0,
```
Comment (ratankaliani) [inline script/src/input.rs]: nit: Prob better practice to put this on it's own line, not in the return struct.

File: rotate/src/main.rs [hunk 10]
```diff
@@ -19,28 +17,14 @@ pub fn main() {
 
     // Compute new authority set hash & convert it from binary to bytes32 for the blockchain
     let new_authority_set_hash: Vec<u8> =
-        compute_authority_set_commitment(justification.num_authorities, justification.pubkeys);
+        compute_authority_set_commitment(justification.num_authorities, justification.pubkeys.clone());
```
Comment (ratankaliani) [inline rotate/src/main.rs]: Fix: These should use `num_authorities` and `pubkeys` from `header_rotate_data`.

File: script/src/bin/rotate.rs [hunk 11]
```diff
@@ -53,4 +57,40 @@ async fn main() {
...
+#[cfg(test)]
+mod tests {
+    use sp1_vectorx_primitives::compute_authority_set_commitment;
+
+    use super::*;
+
+    #[tokio::test]
+    #[cfg_attr(feature = "ci", ignore)]
+    async fn test_compute_authority_set_commitment() {
+        // Generate & proof as usual.
```
Comment (ratankaliani) [inline script/src/bin/rotate.rs]: nit: Remove

**Review by ratankaliani — APPROVED** [review]
LGTM, one small nit.

**Pull request merged**